### PR TITLE
States data

### PR DESCRIPTION
**Pull request by mcsloy**
This pull request introduces the state entities and their factory subroutines.

### CODE REVIEW

File: src/data.jl [hunk 1]
```diff
@@ -0,0 +1,1159 @@
...
+    blk_idxs = filter_on_site_idxs(blk_idxs)
+    blk_idxs = repeat_atomic_blk_idxs(blk_idxs, 1)
+    return get_sub_blocks(matrix, blk_idxs, s_i, s_j, atoms, basis_def), blk_idxs
+end
+
+
+
+end
+
+module States
```
Comment (jameskermode) [inline src/data.jl:667]: long file -maybe move into States.jl and include?

File: test/unit_tests/test_data.jl [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+function _error_free(expr::Expr)
+    try
+        eval(expr)
+        return true
+    catch
+        return false
+    end
+end
+
+@testset "States" begin
```
Comment (jameskermode) [inline test/unit_tests/test_data.jl:515]: could be split to test_staes.jl

**Review by jameskermode — APPROVED** [review]

File: src/data.jl (hunk 1 above)
Comment (mcsloy) [reply src/data.jl:667]: Agreed.

File: test/unit_tests/test_data.jl (hunk 2 above)
Comment (mcsloy) [reply test/unit_tests/test_data.jl:515]: Agreed.

**Pull request merged**